lazer/sdk/rust/client/src/history_client.rs

```diff
@@ -0,0 +1,401 @@
+            .cache_dir
+            .as_ref()
+            .map(|path| path.join("symbols_v1.json"))
+    }
+
+    /// Fetch current metadata for all symbols.
+    pub async fn all_symbols_metadata(&self) -> anyhow::Result<Vec<SymbolMetadata>> {
+        self.fetch_symbols_initial().await
+    }
+
+    /// Fetch metadata for all symbols as an auto-updating handle.
+    ///
+    /// Returns an error if the initial fetch failed.
+    /// The returned `SymbolMetadataHandle` will be updated by a background task when the data changes.
+    pub async fn all_symbols_metadata_handle(&self) -> anyhow::Result<SymbolMetadataHandle> {
+        let symbols = Arc::new(
+            self.fetch_symbols_initial()
+                .await?
+                .into_iter()
+                .map(|f| (f.pyth_lazer_id, f))
+                .collect::<HashMap<_, _>>(),
+        );
+        let previous_symbols = symbols.clone();
+        let handle = Arc::new(ArcSwap::new(symbols));
+        let client = self.clone();
+        let weak_handle = Arc::downgrade(&handle);
+        tokio::spawn(async move {
+            client
+                .update_symbols_handle(weak_handle, previous_symbols)
+                .await;
+        });
+        Ok(SymbolMetadataHandle(handle))
+    }
+
+    /// Fetch metadata for all symbols as an auto-updating handle.
+    ///
+    /// The returned `SymbolMetadataHandle` will be updated by a background task when the data changes.
+    /// If the initial fetch failed, the handle will initially contain an empty hashmap.
+    pub async fn all_symbols_metadata_fault_tolerant_handle(&self) -> SymbolMetadataHandle {
+        let initial_result = self.fetch_symbols_initial().await;
+        let symbols = match initial_result {
+            Ok(data) => data
+                .into_iter()
+                .map(|f| (f.pyth_lazer_id, f))
+                .collect::<HashMap<_, _>>(),
+            Err(err) => {
+                warn!(
+                    ?err,
+                    "failed to fetch symbols, proceeding with empty symbol list"
+                );
+                HashMap::new()
+            }
+        };
+        let symbols = Arc::new(symbols);
+        let previous_symbols = symbols.clone();
+        let handle = Arc::new(ArcSwap::new(symbols));
+        let weak_handle = Arc::downgrade(&handle);
+        let client = self.clone();
+        tokio::spawn(async move {
+            client
+                .update_symbols_handle(weak_handle, previous_symbols)
+                .await;
+        });
+        SymbolMetadataHandle(handle)
+    }
+
+    /// Fetch metadata for all symbols as a receiver.
+    ///
+    /// Returns an error if the initial fetch failed.
+    /// On a successful return, the channel will always contain the initial data that can be fetched
+    /// immediately from the returned receiver.
+    /// You should continuously poll the receiver to receive updates.
+    ///
+    /// Panics if the buffer capacity is 0.
```

Comment:
since we are returning a `Result` let's return Err instead of panicing.

Reply:
Done.